### PR TITLE
Implement `maxUploadFileSize`

### PR DESCRIPTION
**Pull request by zone117x**
Progress towards [Production ready node.js HTTP server ](https://github.com/blockstack/gaia/issues/253)

Completes issues:
* https://github.com/blockstack/gaia/issues/161 GET /hub_info should advertise the hub's maximum file size 
* https://github.com/blockstack/gaia/issues/187 Edit gaia hub drivers to check content-size on gaia hub writes 

When a content-length header is provided (almost always the case) then it is checked against the hub's max configured size. If the size is too large, the http request is immediately rejected with a specific error for the client (blockstack.js) to be able to handle appropriately. 

Additionally, while the a request payload is being streamed, the size is monitored and if the content-length is exceeded, the stream is immediately destroyed. 

If a content-length header is not provided, then the max hub size limit is used instead. This is to allow for clients that do not necessarily know the content-length in advance. The `fetch` w3c spec specifies that a ReadableStream (which has an unknown length) should be allowed as the POST body payload. This is a feature we may want to use in the future for streaming encryption and network request. 

## Manual Testing

TODO: Memory load testing during large file writes to ensure there are no performance regressions. 

## Automated Testing

Near full test coverage. SuperAgent is a PITA and difficult to spoof malicious client behavior. Some error paths are not covered in the http code reporting layer, but covered in the error handling layer right beneath it. 

## Submission Checklist

- [x] Based on correct branch: feature submissions should be on `develop`, hotfixes should be on `master`

- [x] The code passes our eslint definitions, unit tests, and
      contains correct TypeFlow annotations.

- [x] Submission contains tests that cover any and all new functionality or code changes.

- [x] Submission documents any new features or endpoints, and describes how developers
      would be expected to interact with them.

- [x] Author has agreed to our contributor's agreement.

### CODE REVIEW

File: hub/src/server/server.ts [hunk 1]
```diff
@@ -240,7 +256,45 @@ export class HubServer {
...
+    const monitorPipeline = pipelineAsync(stream, monitorStream)
+
+    // Create a PassThrough stream to give to driver for uploading to storage backend. 
+    const uploadStream = new PassThrough()
+    const uploadPipeline = pipelineAsync(stream, uploadStream)
+
+    // Now that the PassThrough streams and pipes are setup, the data event listener
+    // for the size monitor can be setup without either stream listeners missing any chunks. 
+    let monitoredContentSize = 0
+    monitorStream.on('data', (chunk: Buffer) => {
```
Comment (zone117x) [inline hub/src/server/server.ts]: Reviewer note:
This creates two reader streams for a file upload. This has the potential, depending on node/V8 optimizations, to briefly duplicate network chunk streams in memory. But in general, this is the recommended approach to properly handling backpressure and monitoring stream progress when there are multiple consumers.

File: hub/src/server/server.ts [hunk 2]
```diff
@@ -240,7 +256,29 @@ export class HubServer {
       }
     }
 
-    const readURL = await this.driver.performWrite(writeCommand)
+    // Create a PassThrough stream to monitor streaming chunk sizes. 
+    const { monitoredStream, pipelinePromise } = monitorStreamProgress(stream, totalBytes => {
+      if (totalBytes > this.maxFileUploadSizeBytes) {
```
Comment (kantai) [inline hub/src/server/server.ts]: This should check against the Content-Length header -- if the uploader is streaming more than the claimed Content-Length, we could close and error.

Comment (zone117x) [reply hub/src/server/server.ts]: Good catch, fixed.

**Comment by jeffdomke** [comment]
Sounds like next steps are @jcnelson finishes review then @timstackblock will help on the memory  load testing.

File: hub/src/server/server.ts [hunk 3]
```diff
@@ -208,14 +213,25 @@ export class HubServer {
...
-      path, stream, contentType,
-      contentLength: parseInt(requestHeaders['content-length'] as string)
+    await this.proofChecker.checkProofs(address, path, this.getReadURLPrefix())
+
+    const contentLengthHeader = requestHeaders['content-length'] as string
+    const contentLengthBytes = parseInt(contentLengthHeader)
+    const isLengthFinite = Number.isFinite(contentLengthBytes)
+
+    if (!isLengthFinite) {
+      const errMsg = `A valid 'Content-Length' header must be passed. Received header "${contentLengthHeader}"`
```
Comment (jcnelson) [inline hub/src/server/server.ts]: I think we should mention somewhere in the README that a `Content-Length` header is necessary -- there's no support for chunked encoding.

File: hub/src/server/utils.ts [hunk 4]
```diff
@@ -43,6 +43,14 @@ export function getDriverClass(driver: DriverName): DriverConstructor & DriverSt
   }
 }
 
+export function megabytesToBytes(megabytes: number) {
+  return megabytes * 1024 * 1024
+}
+
+export function bytesToMegabytes(bytes: number, decimals = 2) {
+  return Number.parseFloat((bytes / 1024 / 1024).toFixed(decimals))
```
Comment (jcnelson) [inline hub/src/server/utils.ts]: Not a bug, but I'd prefer `bytes / (1024 * 1024)` -- I think `/` is always left-binding, but this would play it safe.

**Comment by timstackblock** [comment]
Confirmed the hub and reader is successfully running. Overall, the PR respects upper file limits but doesn’t respect lower limits. 

	•	3 sample files are needed: 
             file1 is less than 100KB, 
             file2 is greater than 100KB but less than upperlimit (20MB) default and 
             file3 is greater than upperlimit.

File1 upload (Successful):
	•	File upload was successful despite being less than the lower limit of 100KB.
 
File2 Upload(Successful):
	•	File upload was successful given it was within the range of acceptable file size limit.
 
File3 Upload(Unsuccessful):
	•	File size exceeds upper size limit.
	•	Response: "message":"Max file upload size is 20 megabytes. Rejected Content-Length of 109.4165 megabytes

File: hub/src/server/server.ts (hunk 3 above)
Comment (zone117x) [reply hub/src/server/server.ts]: Per my comment https://github.com/blockstack/gaia/pull/256#issuecomment-525489836
I removed the requirement for providing a content-length header. After extensive testing, this is a frustrating header to provide given CORs whitelist contraints. Additionally, it excludes future uploading features that do not necessarily know the content length in advance.

File: hub/src/server/utils.ts (hunk 4 above)
Comment (zone117x) [reply hub/src/server/utils.ts]: Fixed

**Comment by zone117x** [comment]
I removed the requirement for providing a content-length header. After extensive testing, this is a frustrating header to provide given CORs whitelist contraints. Additionally, it excludes future uploading features that do not necessarily know the content length in advance. 

If the header is provided by a well-behaving and non-streaming hub and client, then it results in nice immediate errors immediately provided to the client. If the header is not provided -- as is currently the case with the current PBC hub -- then the error is provided to the client after the hub-configured limit has been exceeded. 

I have pending future PRs that increase the probability of Gaia hub servers and clients reporting the content-length values. However, the possibility of the header not being provided is a situation in which I think we should reasonably expect and allow for. 

@jcnelson @kantai can you re-review? Thanks.

**Pull request merged**